core/store/src/trie/shard_tries.rs

```diff
@@ -156,6 +219,28 @@ impl ShardTries {
         self.get_trie_for_shard_internal(shard_uid, state_root, false, None)
     }
 
+    pub fn get_trie_with_block_hash_for_shard_from_snapshot(
+        &self,
+        shard_uid: ShardUId,
+        state_root: StateRoot,
+        block_hash: &CryptoHash,
+    ) -> Result<Trie, anyhow::Error> {
```

Comment:
From what I understand `anyhow::Error` is generic error, but we already have `StorageError` for storage-related code, I think `StorageInconsistentState` case fits here and below.